### ISSUE TITLE
Tentative Roadmap.

### ISSUE THREAD

**Issue by vpenades** [vpenades/SharpGLTF.Monogame.Example#3]
This repository served originally as a showcase of how to use SharpGLTF to load a 3D model into a game engine, in this case, MonoGame.

Since then, it's been evolving, and given that MonoGame is severely lacking a decent 3D API, we're trying to provide an alternative for the in-built monogame 3D model features.

### Roadmap
- Improve PBR shaders support.
- Refactor runtime model classes into it's own library, so it no longer depends on glTF, so other formats my be used.
  - Copy and refactor some runtime model classes from glTF to this project.
  - Provide a solution for animation curve sampling. (not dependant on glTF).
- Try other platforms (UWP, Android, etc)
- Consider content pipeline processing.

### Content Pipeline processing.

The two reasons why this library loads glTF models at runtime instead of going through the content pipeline is because:
- The runtime requires a whole new set of 3D model related supporting classes.
- The content pipeline is constrained to a particular, and outdated, 3D model architecture.

Right now, improving the content pipeline is a dauntling tasks that nobody seems to be willing to trackle, and it would probably cause lots of backwards compatibility issues. Also, 3D models format technology is continuosly evolving, so there's no point in doing a lot of work, just to become outdated after a few years.

So my proposal is to _take away 3d model processing from the main pipeline_ and make it a separated tool. Pretty much like the effects compiler tool. It could be made so it can be called by MGCB seamlessly, but it would no longer be integrated within the main library.

In order to do this stand alone 3D model pipeline tool, there's a prerequirement: the whole set of class serialization writing needs to be exposed, so a tool can write its own XNBs from scratch.

**Comment by NicusorN5** [comment]
Providing some arguments to the features that have a question mark:

1.) I'm my opinion, shadow mapping should be implemented by the user itself, since BasicEffect Model instances may be used for really basic objects like props, map elements, etc. (Since shadow mapping requires redrawing the scene, playing with the stencil buffer, and some more stuff).
2.) For the model picking - I have a Model-Ray intersection algorithm written in C# Monogame (Moller-Trumbore to be exact) written on an older push in my game engine's code. Check that out here: https://github.com/3DRadSpace/3D_Rad_Space/blob/7feb0daa41190d244bfab2d109fd62f97dadba19/3DRadSpace/3DRadSpace/Editor.cs#L348
Feel free to adapt it to your own code.
3.) MipMap generation and texture compression sound good. Why not?

**Comment by vpenades** [comment]
@NicusorN5 Ideally, the ultimate objective is to have a framework than can be used, _out of the box_ to achieve similar quality results compared to the basic implementation of Unity.

There's a good number of developers that don't really wish to have to develop their own shaders on every project they start, and need stuff that just works. This is what I think monogame needs to keep being competitive.

This doesn't mean I am going to do all the stuff listed in the roadmap... it's mostly a _wishlist_ of what I would like this project to solve... and I would like people to contribute solutions for some of these issues.


> 1.) shadow mapping...
This is a wishlist of mine, and I don't hope to be able to do it myself... it's in the category of things I would like to have in monogame out of the box.

> 2.) For the model picking
I was already thinking on solutions for picking, since it's something I've already done in the past... but I will certainly look at your solution, thanks!

> 3.) MipMap generation
This is a very much needed feature required by glTF to look good on monoGame... but it's also a feature that needs to be implemented by monogame straight away. This includes not only supporting mipmap generation, but also supporting newer texture formats like WEBP and KTX2, which are supported by glTF.

**Comment by NicusorN5** [comment]
Coming back, I've found some articles regarding implementing shadow mapping in XNA. They are actually published again, in a GitHub Wiki for a while, and I've found out that there's a very small number of lines to change (change the shader version). 

https://github.com/SimonDarksideJ/XNAGameStudio/wiki/Riemers3DXNA3hlsl12shadowmap

You might have seen this, but I think this might be helpful, anyways.

**Comment by vpenades** [comment]
> You might have seen this, but I think this might be helpful, anyways.

No, I didn't... and the shader code provided by the link is short enough to give it a try at some point.
